### PR TITLE
Cirrus: Collect benchmarks on machine instances

### PR DESCRIPTION
**Pull request by cevich**
The hardware used for podman-machine testing is fairly stable/predictable because it's bare-metal.  This is a nearly ideal environment for collection of benchmarking data.  Arrange for that to happen, and the resulting data to be collected.

Signed-off-by: Chris Evich <cevich@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

File: contrib/cirrus/lib.sh
Comment (vrothberg) [inline contrib/cirrus/lib.sh:320]: @edsantiago may have some ideas for that.

Comment (cevich) [reply contrib/cirrus/lib.sh:320]: Oh, that was just a note to myself.  I plan on storing all the goodies in a cirrus artifact.

**Comment by cevich** [comment]
@edsantiago @vrothberg we've got some [data being generated now (example)](https://api.cirrus-ci.com/v1/artifact/task/4534778473480192/benchmark/data/benchmarks.txt).  However it doesn't seem very useful in that format.  Does/can that `localbenchmarks` target output to a file with the data so we don't need to scrape stdout?  Thoughts?

**Comment by edsantiago** [comment]
https://github.com/containers/podman/blob/6c37cf332439989148e408af49afaf7c0cb6e166/hack/parse-localbenchmarks

**Comment by cevich** [comment]
That's an option we could use.  I'm having "how hard could it be" thoughts of outputting right from golang into a file/files.  OTOH, I don't want to disturb the measurements by doing that.

Also..."how hard could it be" often guarantees weeks of work and bickering over code quality.  Maybe better to just use @edsantiago perl that only he can read anyway :rofl:

I guess my main motivation for wanting to craft the output from the source is my golang-interest (a.k.a. for the funzies).  It would also be easier for a wider-community of developers to access/change/modify/fix/etc.

Unless there are strong opinions one way or another, I'll probably ponder this for a few days, then throw up my hands and just use the perl.

**Comment by cevich** [comment]
@edsantiago help please (when you get a chance).  The [raw benchmarks output artifact](https://cirrus-ci.com/task/6036788599324672) (task link) looks right, but the parsed one is definitely wrong.  [Am I running it wrong](https://github.com/containers/podman/pull/16449/files#diff-87b53ab90b08478b1624b8f218f6d0b6eab834f2cf81dd73581540c3a0dfb63cR324-R325)?

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:325]: Missing `<`. It's a filter. 
```bash
hack/... < $datadir/...
         ^-------
```

Comment (cevich) [reply contrib/cirrus/lib.sh:325]: Ahh that would explain it.  Thanks!

**Comment by cevich** [comment]
@vrothberg & @edsantiago PTAL this is ready to go.  You'll find the benchmarks.csv and environment file are both now included in [the artifacts archive.](https://cirrus-ci.com/task/5442393579716608).  This should make them easy to retrieve by secondary tooling per-task, and for the latest run on any branch via the artifacts Cirrus-API.

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:320]: Does this really have to be so complicated? `grep -ce '^processor' /proc/cpuinfo` is instantly grokkable, no brain overheating needed.

Also, these don't need to be spelled out as CPUTOTAL MEMTOTAL ETC. It might be nicer for readers to just have them spelled out: `cpu total:`, `Mem Total`

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:318]: You know how I feel about duplication. (Prime example right here, where one envariable is misleadingly called the wrong name). Since `$SECRET_ENV_RE` is so nicely defined above, could you instead do
```bash
    env | grep -Ev "$SECRET_ENV_RE" | sort > $datadir/benchmarks.env
```

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:321]: Might be more readable: `awk -F: '$1 == "MemTotal" { print $2 }' </proc/meminfo`

**Review by edsantiago — COMMENTED** [review]
Rebase should be easy. A few suggestions for maintainability.

File: contrib/cirrus/lib.sh
Comment (cevich) [reply contrib/cirrus/lib.sh:320]: Oh, `grep -c` will you look at that!  Didn't know that existed, thanks!

> These don't need to be spelled out as

This file is intended to be machine-readable metadata details indicating the basis for the benchmarks.  It's not needed now, but immensely useful if somebody wants to do any trend analysis or averaging over long periods of time.  Then, they will need a machine-readable way to say "These, ran under the exact same conditions, as those [benchmarks]".

I can add a comment-line at the top of the file so there's no confusion.

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:316]: Missing `/proc/cpuinfo`. And, I think you missed my comment about duplication.

File: contrib/cirrus/lib.sh
Comment (cevich) [reply contrib/cirrus/lib.sh]: > one envariable is misleadingly called the wrong name

`INST_TYPE` was done on purpose since it's not the consumers job to pick apart cloud-differences.  The only reason we don't use an `INST_TYPE` globally is Cirrus-CI didn't support it initially.  Now it does, and *someone* (me) might decide to go fix it one day :wink: 

I can do the sorting and filtering though, that's a no-brainier.

File: contrib/cirrus/lib.sh
Comment (cevich) [reply contrib/cirrus/lib.sh:316]: oh oops.  Yeah, my carlessness-o-meter is maxed out today, sorry.  Only cure is more sleep :disappointed:

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:306]: My request was `printenv | grep -Ev`, with the goal being to eliminate duplication. The grep added here is pretty much a NOP, and a misleading/confusing one.

(I don't think it's what's causing the build failure, though. I don't know what the problem is there).

Comment (cevich) [reply contrib/cirrus/lib.sh]: ...okay, I think I got it now.

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:308]: This is where the grep-sort belongs, not below:
```bash
printenv | grep -Ev "$SECRET_ENV_RE" | sort
```

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:317]: This is un-ideal. It will mingle the non-environment with real-environment, making it hard for a future reader. See above. Sorry for not having been clearer.

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:310]: This would be a great place to add something like
```
# these are not actual environment variables; all of these are included as a representation of system configuration
```

**Review by edsantiago — COMMENTED** [review]
CI is still broken. A few suggestions.

File: contrib/cirrus/lib.sh
Comment (cevich) [reply contrib/cirrus/lib.sh:310]: Yeah, that's what I was trying to say in the comments.  K.  Let me tweak it some more.

**Comment by cevich** [comment]
> Out of curiosity, why do we use --color=always?

oh, it's a copy-paste from what I have in a `.pre-commit-config.yaml`, I can kill it.

**Comment by cevich** [comment]
okay...once again for the fourteenth time...please behave shellcheck...

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:304]: I have a 75% suspicion that you intended this for the `.env` file. Doesn't seem important enough to re-push for...

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:317]: ...but unfortunately this is. Should be `>>`, not `>`.

**Review by edsantiago — COMMENTED** [review]
Sorry.

**Comment by edsantiago** [comment]
Don't re-push yet, though: let's see if CI passes. I don't see a fix for the CI-break from a few pushes ago, so it might still need more fixes.

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:318]: It seems likely that the failure is here, and `&>` is hiding the message.

Comment (cevich) [reply contrib/cirrus/lib.sh:318]: Right, and the data collection doesn't happen if the job fails so we can't see it.  Let me fix that in `.cirrus.yml` so we always collect the data, success or not.

Comment (cevich) [reply contrib/cirrus/lib.sh:318]: I'll also change that to a `) |& tee $datadir/benchmarks.raw`

Comment (cevich) [reply contrib/cirrus/lib.sh:318]: rather `make localbenchmarks |& tee $datadir/benchmarks.raw`

File: contrib/cirrus/lib.sh
Comment (edsantiago) [inline contrib/cirrus/lib.sh:318]: Stupid question maybe, but I'm ultra-confused. Why stderr? I mean, what possible use is it in `benchmarks.raw`? For me the solution was:
```diff
- make ... &>
+ make ... >
```
Once we figure out the failure problem, could you please just simplify this to a simple `>`?

Comment (cevich) [reply contrib/cirrus/lib.sh:318]: oh, this is just a leftover from the original implementation, where I didn't know how the benchmarks were reported (stdout or stderr).  Yep, NP fixing, I'll queue it up now.

**Comment by edsantiago** [comment]
Aha, thanks. This is the error:
```
/bin/sh: line 1: ./test/tools/build/ginkgo: No such file or directory
```
This seems like one way to fix it:
```diff
diff --git a/Makefile b/Makefile
index 71c0d045f..9241d3071 100644
--- a/Makefile
+++ b/Makefile
@@ -581,7 +581,7 @@ localmachine: test-binaries
 	$(MAKE) ginkgo-run GINKGONODES=1 GINKGOWHAT=pkg/machine/e2e/. HACK=
 
 .PHONY: localbenchmarks
-localbenchmarks: test-binaries
+localbenchmarks: test-binaries .install.ginkgo
 	PATH=$(PATH):$(shell pwd)/hack ACK_GINKGO_RC=true $(GINKGO) \
 		      -focus "Podman Benchmark Suite" \
 		      -tags "$(BUILDTAGS) benchmarks" -noColor \
```

**Comment by cevich** [comment]
> /bin/sh: line 1: ./test/tools/build/ginkgo: No such file or directory

Ugh, no won't work, the binary moved.  I thought I fixed that problem in another PR...

**Comment by cevich** [comment]
Oh bother!  I am not having a smooth week :disappointed:

**Comment by edsantiago** [comment]
Still failing the same way:
```
/bin/sh: line 1: ./test/tools/build/ginkgo: No such file or directory
```
I'm stumped. Only thing I can think of that will help is `set -x`.

Artifacts uploaded, though, and oops, the `sort` was a bad idea. Sorry. It needs to be removed.

**Comment by cevich** [comment]
argh, what a hell of a PR.  I may save this one for next year, a fresh mind, and `get_ci_vm.sh` :grinning:

**Comment by cevich** [comment]
FFS it's finally working.  And collecting properly formatted CSV data (thanks @edsantiago !).

**Comment by cevich** [comment]
status: This is waiting for final review/approval of Ed (who's on PTO).

**Review by TomSweeneyRedHat — COMMENTED** [review]
LGTM
but definitely want @edsantiago 's eyeballs on this.

**Comment by edsantiago** [comment]
LGTM but this is too old and scary to merge now. I'll slash-lgtm it in the morning, when I'll be around to monitor for problems.

**Comment by edsantiago** [comment]
gulp

/lgtm

**Review by edsantiago — APPROVED** [review]

**Pull request merged**
